tokio/src/sync/rwlock.rs

```diff
@@ -0,0 +1,151 @@
+//! An asynchronous `RwLock`-like type.
+//!
+//! This module provides [`RwLock`], a type that acts similarly to an asynchronous `RwLock`.
+//!
+//! This allows you to do something along the lines of:
+//!
+//! ```rust,no_run
+//! use tokio::sync::RwLock;
+//! use std::sync::Arc;
+//!
+//! #[tokio::main]
+//! async fn main() {
+//!     let data1 = Arc::new(RwLock::new(0));
+//!     let data2 = data1.clone();
+//!
+//!     tokio::spawn(async move {
+//!         let mut lock = data2.write().await;
+//!         *lock += 1;
+//!     });
+//!
+//!     let mut lock = data1.write().await;
+//!     *lock += 1;
+//! }
+//! ```
+//!
+//! [`RwLock`]: struct.RwLock.html
+//! [`RwLockReadGuard`]: struct.RwLockReadGuard.html
+//! [`RwLockWriteGuard`]: struct.RwLockWriteGuard.html
+
+use crate::sync::semaphore;
+use crate::sync::Mutex;
+use futures_util::future::poll_fn;
+use std::cell::UnsafeCell;
+use std::ops;
+
+const MAX_READS: usize = 32;
+/// An asynchronous readers writer lock usefull for protecting shared data
+///
+/// Each lock has a type parameter (`T`) which represents the data that it is protecting. The data
+/// can only be accessed through the RAII guards returned from `read` and `write`, which
+/// guarantees that the data is only ever modified when the rwlock is with write lock.
```

Comment:
Check the grammar on that last sentence.  Might I suggest "only ever modified when the rwlock is locked in write mode".